tmt/beakerlib.py

```diff
@@ -222,27 +249,75 @@ def fetch(self) -> None:
             # Clone repo with disabled prompt to ignore missing/private repos
             try:
                 if self.url:
+                    global DEFAULT_CLONE_DIR
+                    if not DEFAULT_CLONE_DIR:
+                        DEFAULT_CLONE_DIR = os.path.join(
+                            self.parent.workdir, DEFAULT_DESTINATION, 'clone')
```

Comment:
`DEFAULT_CLONE_DIR` is obviously not a module-wide default, being composed of one very specific `parent`'s workdir - would it make sense to store this value as a `Run` or `Plan` attribute?

There is a single place where `parent` is defined, and that's `discover/fmf` when calling `tmt.beakerlib.dependencies()`:

```python
                test.require, test.recommend, _ = tmt.beakerlib.dependencies(
                    test.require, test.recommend, parent=self)
```

`parent`'s then always (at this moment) `DiscoverFmf`, so let's say it's a subclass of `DiscoverPlugin` - how about making that official, let `parent` always be `DiscoverPlugin` or its subclass, and use it - or its `step.plan` - to hold the "default for this plan" directory for beaker libraries? Or `step.plan.my_run` if such storage is rather per-run business. I can imagine `Plan` or `Run` gaining new property, let's say `beakerlib_clone_dirpath`, returning always `os.path.join(self.workdir, 'libs', 'clone')` - Beakerlib support is fairly significant, there's no shame in the support "sneaking" into core classes in a nicely defined & documented way. I suppose the support for multiple `discover` steps, sharing the library storage, which rules out adding such a property to the `Library` class itself as the `parent` might be different every time - if that's the case, `Plan` or `Run` holding the value would make sense to me. I'd consider it much better that a global not-so-default.

The thing is, this `DEFAULT_CLONE_DIR` is a new shared mutable state, and with prospects of parallelization, there may be a day when different parts of tmt would have different ideas about what this **default** is supposed to be. I'd very much like to avoid such a day, by not even getting tempted by the idea. If we can clearly define the scope of this storage directory, let's do it now and here, when it's being added.

Reply:
Thanks for pointing out the obvious solution that I missed, although am only going with having it under DiscoverFmf as that is IMO the only relevant place for it, not sure how multiple discover steps will work so don't want to work with that.

Otherwise I want that location relevant for single plan, multiple plans (and parallelisation) might clash if I was to place it 'higher', at least with current implementation, so either way it will require some attention when relevant code is added.

Reply:
On second thought, let's have it in Common as typing was not happy and it makes sense there as well.

Reply:
No problem, glad to help & not be thrown out :)

I'd probably not put this into `Common` as that class is used as base for many different classes, many of those never heard of any Beakerlib (and never will...). `DiscoverFmf` or its parent class, `DiscoverPlugin`, seemed best to me - what kind of typing issues did you run into? I suppose it was something related to `self.parent` not having this newly added property - we can ammend that, with our human knowledge of the wider picture, for example `assert isinstance(self.parent, tmt.steps.discover.fmf.DiscoverFmf)` should make mypy satisfied, and it would not be in conflict with how things are *de facto*, if not *de jure*.

Reply:
Interesting, it works with isinstance, one small problem might occur if someone in the future used the Library class without parent argument, but let's care about that when it occurs as now there is only one place this is called from

Reply:
aaaaaand I'm wrong 😄 it's actually used in integration/test tests/test_qe.py::test_tmt_to_wrapper_args with Common class, I can't even find where these tests come from

What if I made it more common as in change the beakerlib_clone_dirpath to more generic one, like 'clone_dirpath' and drop the libs part, then it would make sense in other places as well, for example we could do the same for tests if at some point we decide to only filter tests that we are running

Reply:
No idea about generalization, no preference either, as I don't have any use for it and it seems like tmt's cloning beakerlibs only. It would make beakerlibs less prominent, so very small +1 from me, but people more involved with beakerlibs should be heard, too.

Reply:
This path doesn't really have anything specific to do with beakerlibs as I'm just cloning data there and cherry-picking directories/files I actually need, it is removed at the end of discover step. In reality I could just hardcode it for /tmp/clone and be done with it, but I wanted to keep the clone dir close.

@psss @lukaszachy what do you think about clone_dirpath property for Common class?